src/cycle.rs

```diff
@@ -58,6 +61,52 @@ use crate::key::DatabaseKeyIndex;
 /// Should only be relevant in case of a badly configured cycle recovery.
 pub const MAX_ITERATIONS: u32 = 200;
 
+pub struct UnexpectedCycle(Option<crate::Backtrace>);
+
+impl fmt::Debug for UnexpectedCycle {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("cycle detected but no cycle handler found")?;
+        if let Some(backtrace) = &self.0 {
+            f.write_str(": ")?;
+            backtrace.fmt(f)?;
+        }
+        Ok(())
+    }
+}
+
+impl fmt::Display for UnexpectedCycle {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("cycle detected but no cycle handler found")?;
+        if let Some(backtrace) = &self.0 {
+            f.write_str("\n")?;
+            backtrace.fmt(f)?;
+        }
+        Ok(())
+    }
+}
+
+impl UnexpectedCycle {
+    pub(crate) fn throw() -> ! {
+        // We use resume and not panic here to avoid running the panic
+        // hook (that is, to avoid collecting and printing backtrace).
+        panic::resume_unwind(Box::new(Self(crate::Backtrace::capture())));
+    }
+
+    /// Runs `f`, and catches any salsa cancellation.
```

Comment:
Nit: Comment needs updating (and the code as well) because it doesn't catch cancellation